### PR TITLE
feat(firewood/ffi): add `FjallStore`

### PR DESCRIPTION
**Pull request by RodrigoVillar**
This PR adds `FjallStore`, an implementation of `RootStore` that allows for persistent storage of revisions (in contrast to `MockStore`, which keeps everything in-memory). The main changes of this PR are as follows:

- Adds `FjallStore` implementation
- Extends the `DbConfig` to take in a path to where `FjallStore` will be stored
- Extends the FFI database config to enable archival support
- Adds `FjallStore` tests in both Rust and Golang

### CODE REVIEW

File: firewood/src/root_store.rs [hunk 1]
```diff
@@ -108,3 +109,83 @@ impl RootStore for MockStore {
...
+}
+
+impl Debug for FjallStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("FjallStore")
+            .field("keyspace", &"<Keyspace>")
+            .field("items", &"<PartitionHandle>")
+            .finish()
+    }
+}
```
Comment (RodrigoVillar) [inline firewood/src/root_store.rs]: Since `Keyspace` and `Partition` both don't implement the `Debug` trait, I manually implemented `Debug` for `FjallStore`. Open to any suggestions on how to make this better though.

**Comment by RodrigoVillar** [comment]
Note to reviewers: I've marked the PR as `DO NOT MERGE` since I'm working on getting benchmark results for Firewood + FjallStore when compared to LevelDB (and we should only merge once we have benchmark results). However, I figured we can get the code review process going since benchmarking can take place async.

File: ffi/firewood_test.go [hunk 2]
```diff
@@ -1264,6 +1264,55 @@ func TestHandlesFreeImplicitly(t *testing.T) {
...
+		r.NoError(err)
+	}
+
+	// Close and reopen the database
+	r.NoError(db.Close(t.Context()))
+
+	db, err = New(dbFile, config)
+	r.NoError(err)
+
+	// Verify that we can access all revisions
```
Comment (rkuris) [inline ffi/firewood_test.go:1298]: In theory, we should be able to fix the original issue here and then this test will pass even if fjall is not doing anything. Can we configure the revision manager when creating the database to only save 5 revisions?

Additionally, we should test that all 10 are readable before reopening the database as well...

File: ffi/src/handle.rs [hunk 3]
```diff
@@ -100,9 +113,19 @@ impl DatabaseHandle {
...
+            .map_err(|e| invalid_data(format!("root store path contains invalid utf-8: {e}")))?;
+
+        let root_store_dir = Some(root_store_path)
+            .filter(|s| !s.is_empty())
+            .map(PathBuf::from);
+
         let cfg = DbConfig::builder()
             .truncate(args.truncate)
             .manager(args.as_rev_manager_config()?)
+            .root_store_dir(root_store_dir)
```
Comment (rkuris) [inline ffi/src/handle.rs:128]: this works, but I think we need to switch to a directory per our discussion (probably a different PR).

File: firewood/src/db.rs [hunk 4]
```diff
@@ -122,6 +122,9 @@ pub struct DbConfig {
     // TODO: Experimentally determine the right value for BatchSize.
     #[builder(default = UseParallel::BatchSize(8))]
     pub use_parallel: UseParallel,
+    /// `RootStore` directory path
+    #[builder(default = None)]
```
Comment (rkuris) [inline firewood/src/db.rs:129]: Question: Isn't `None` the default if you don't specify?

File: firewood/src/root_store.rs (hunk 1 above)
Comment (rkuris) [reply firewood/src/root_store.rs]: You can use [derive_where](https://docs.rs/derive-where/latest/derive_where/) with [skip_inner](https://docs.rs/derive-where/latest/derive_where/#skipping-fields).

File: firewood/src/root_store.rs [hunk 5]
```diff
@@ -108,3 +109,83 @@ impl RootStore for MockStore {
...
+    /// # Errors
+    ///
+    /// Will return an error if unable to create or open an instance of `FjallStore`.
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<FjallStore, RootStoreError> {
+        let keyspace = Config::new(path).open().map_err(|e| RootStoreError {
+            method: RootStoreMethod::New,
+            source: Box::new(e),
+        })?;
+        let items = keyspace
+            .open_partition("firewood", PartitionCreateOptions::default())
```
Comment (rkuris) [inline firewood/src/root_store.rs]: Let's pull "firewood" out as a const for better readability as to what this is used for:
```
const FJALL_PARTITION_NAME: &str = "firewood";
```

File: firewood/src/root_store.rs [hunk 6]
```diff
@@ -108,3 +109,83 @@ impl RootStore for MockStore {
...
+            .map_err(|e| RootStoreError {
+                method: RootStoreMethod::New,
```
Comment (rkuris) [inline firewood/src/root_store.rs]: nit: in looking at this, I'm not convinced having the method here adds a ton of value and it increases the amount of code significantly. Your call though, I'm okay either way.

`let items = keyspace.open_partition(...)?;` is soo much easier to read.

File: storage/src/trie_hash.rs [hunk 7]
```diff
@@ -105,6 +105,12 @@ impl TrieHash {
         self
     }
 
+    /// Converts `TrieHash` to an array of bytes.
+    #[must_use]
+    pub const fn to_bytes(&self) -> [u8; TRIE_HASH_LEN] {
+        self.0
```
Comment (rkuris) [inline storage/src/trie_hash.rs]: This feels like `impl Deref`, although perhaps you need it in const context in which case you can ignore this comment.

File: firewood/src/db.rs (hunk 4 above)
Comment (RodrigoVillar) [reply firewood/src/db.rs:129]: No; by removing the attribute here, we'll need to always call `root_store_dir(None)` on the builder whenever we have an instance of the `Db` builder. We also have similar attributes for `create_if_missing` and `truncate`, which are booleans (and should therefore be `false` by default):

https://github.com/ava-labs/firewood/blob/e775eeb178b36fdfb808540b59a4d91f6cc2dac6/firewood/src/db.rs#L110-L116

File: ffi/src/handle.rs (hunk 3 above)
Comment (RodrigoVillar) [reply ffi/src/handle.rs:128]: Noted here: https://github.com/ava-labs/firewood/issues/1393#issuecomment-3504342934

File: firewood/src/root_store.rs (hunk 6 above)
Comment (RodrigoVillar) [reply firewood/src/root_store.rs]: Do you mean the `method` field of `RootStoreError`?

File: firewood/src/root_store.rs (hunk 1 above)
Comment (RodrigoVillar) [reply firewood/src/root_store.rs]: Ah thank you! Switched to using `derive_where` here: [9e3b9e2](https://github.com/ava-labs/firewood/pull/1395/commits/9e3b9e2d61858cf9fe7f59b15f028a704c6d5623)

File: storage/src/trie_hash.rs (hunk 7 above)
Comment (RodrigoVillar) [reply storage/src/trie_hash.rs]: Yup I could use the existing `impl Deref` - done: 2691ac6

File: firewood/src/root_store.rs (hunk 5 above)
Comment (RodrigoVillar) [reply firewood/src/root_store.rs]: Done: 9d40bef

File: firewood/src/root_store.rs (hunk 6 above)
Comment (rkuris) [reply firewood/src/root_store.rs]: yes

Comment (rkuris) [reply firewood/src/root_store.rs]: If you really really want to keep `method` then you could create a specialized function that works like `map_err` but removes the boilerplate. Then the code would be something like:

```
let items = keyspace
            .open_partition("firewood", PartitionCreateOptions::default())
            .map_err(wrap_with(RootStoreMethod::New))?;
```

This is still better as you avoid all the closure cognitive load.

Comment (RodrigoVillar) [reply firewood/src/root_store.rs]: Assuming that we can figure out which method returned the error, I think it's better to remove the `RootStoreError` type altogether - PR which does that can be found here: https://github.com/ava-labs/firewood/pull/1446

Comment (RodrigoVillar) [reply firewood/src/root_store.rs]: Resolving this as #1446 was merged in.

File: ffi/firewood_test.go (hunk 2 above)
Comment (RodrigoVillar) [reply ffi/firewood_test.go:1298]: Done here: [aa64746](https://github.com/ava-labs/firewood/pull/1395/commits/aa64746278d87c8a19d9e167a676389c740ea6f4).

Modifying the database to save only five revisions revealed an edge case that we didn't test - what happens to the oldest revision when we hit our `max_revisions` limit? It turns out that the oldest revision's nodes were reaped, resulting in `TestFjallStore()` failing. 

To fix this, I modified the `commit()` method so that if the oldest revision is referenced by `RootStore`, it's still removed from memory but its nodes are no longer reaped. I also added an additional test `test_fjall_store_with_capped_max_revisions()` on the Rust side to verify that this is indeed true now.

Comment (RodrigoVillar) [reply ffi/firewood_test.go:1298]: Hmm on second thought, I wonder if it's reasonable to just add a `root_store_enabled` bool to the `RevisionManager` and add a revision's nodes to the free list only if `root_store_enabled == true`. In my mind, a revision being referenced by `RootStore` is (basically) logically equivalent to root store being enabled - this would save us from having to make a read to `RootStore` every time we commit.

File: firewood/src/manager.rs [hunk 8]
```diff
@@ -167,7 +167,7 @@ impl RevisionManager {
     ///    The proposal's parent must be the last committed revision, otherwise the commit fails.
     ///    It only contains the address of the nodes that are deleted, which should be very small.
     /// 2. Revision reaping. If more than the maximum number of revisions are kept in memory, the
-    ///    oldest revision is reaped.
+    ///    oldest revision is reaped if not referenced by `RootStore`.
```
Comment (rkuris) [inline firewood/src/manager.rs]: This comment is confusing to me. It's not clear by what "referenced by `RootStore`" means.

What I'm reading this as is "if RootStore knows about this revision, then don't reap it" which isn't what you want. Reaping here refers to removing the revision from memory, which we want.

I'm also worried about when we don't have a RootStore. Aren't all revisions "not referenced by `RootStore`" in that case? Which implies we're never reaping any revisions, which clearly is wrong.

File: firewood/src/manager.rs [hunk 9]
```diff
@@ -76,7 +76,7 @@ pub(crate) struct RevisionManager {
     /// stored in the filebacked storage.
     historical: RwLock<VecDeque<CommittedRevision>>,
     proposals: Mutex<Vec<ProposedRevision>>,
-    // committing_proposals: VecDeque<Arc<ProposedImmutable>>,
+    /// The list of revisions that are on disk by hash.
```
Comment (rkuris) [inline firewood/src/manager.rs]: This comment will become incorrect when deferred persistence is implemented. I think it's self-documenting already so the doc comment doesn't add much value.

The `committing_proposals` thing is dead code that should probably go away.

File: firewood/src/manager.rs [hunk 10]
```diff
@@ -199,27 +200,40 @@ impl RevisionManager {
...
-            if let Some(oldest_hash) = oldest.root_hash().or_default_root_hash() {
-                self.by_hash
-                    .write()
-                    .expect("poisoned lock")
-                    .remove(&oldest_hash);
-            }
+            let oldest_hash = oldest
+                .root_hash()
+                .or_default_root_hash()
+                .expect("root hash should not be none");
```
Comment (rkuris) [inline firewood/src/manager.rs]: I'm not sure this is true. If you delete all keys and are not using eth compatible mode, this will be None, which will crash.

Interesting that this is not tested.

File: firewood/src/manager.rs [hunk 11]
```diff
@@ -339,14 +353,29 @@ impl RevisionManager {
     }
 
     pub fn revision(&self, root_hash: HashKey) -> Result<CommittedRevision, RevisionManagerError> {
-        self.by_hash
+        if let Some(revision) = self
```
Comment (rkuris) [inline firewood/src/manager.rs:340]: This multi-step flow could use some block level comments.

```suggestion
        // First, check the in-memory revision manager
        if let Some(revision) = self
```

File: firewood/src/manager.rs [hunk 12]
```diff
@@ -339,14 +353,29 @@ impl RevisionManager {
...
+            .root_store
+            .get(&root_hash)
+            .map_err(RevisionManagerError::RootStoreError)?
             .ok_or(RevisionManagerError::RevisionNotFound {
-                provided: root_hash,
-            })
+                provided: root_hash.clone(),
+            })?;
+
+        Ok(Arc::new(NodeStore::with_root(
```
Comment (rkuris) [inline firewood/src/manager.rs]: If you have 2 requests for the same revision, they will not share the NodeStore, and instead each one will get a separate nodestore with a reference count of 1.

One way to address this is to keep a cache of existing created NodeStores by hash. This might be a use case for weak arcs so those NodeStores will be dropped when unreferenced.

Maybe something like this:

In `RevisionManager`:
```rust
    by_rootstore: Mutex<HashMap<TrieHash, Weak<NodeStore<Committed, FileBacked>>>>,
```

Then, when looking for a revision, you lock, check, and populate that map (sloppy locking and unwraps to show the idea here):

```rust
    let nodestore_entry = self.by_rootstore.lock().unwrap().entry(&root_hash);
    let result = match nodestore_entry {
        Occupied(entry) => Some(entry.upgrade().unwrap_or_else(rootstore_lookup(&root_hash).unwrap())),
        Vacant(entry) => {
            if let Some(ns) = rootstore_lookup(&root_hash) {
                entry.insert(ns);
                Some(ns)
            } else {
                None
            }
        }
    }
```

This leaves you with the problem of removing stale hashmap entries that point to weak references that are deallocated, which can be solved by cleaning them up. Since you'll be traversing quite often, and the list will be small, maybe you avoid a hashmap and just walk a list looking for the hash you're interested in and dropping anything you find that's stale along the way.

**Review by rkuris — CHANGES_REQUESTED** [review]
I'm not sure I love changing the reaping logic to fix this issue. The commit should have the ability to reap any revision at any time.

I think there are a few possible approaches here:

 1 - Part of reaping is adding it to the `RootStore`
 2 - Every commit adds to `RootStore` so entries are both in the RootStore and in the revision manager (preferred).

File: ffi/firewood_test.go (hunk 2 above)
Comment (rkuris) [reply ffi/firewood_test.go:1298]: Rather than using a bool, let's just use the rootstore that doesn't store anything. This way we don't have to sprinkle "if" statements in various places -- we always call the appropriate functions for the rootstore and the rootstore can choose to do what it wants, including nothing.

**Comment by rkuris** [comment]
> Note to reviewers: I've marked the PR as `DO NOT MERGE` since I'm working on getting benchmark results for Firewood + FjallStore when compared to LevelDB (and we should only merge once we have benchmark results). However, I figured we can get the code review process going since benchmarking can take place async.

I think we determined that this PR does not affect performance, so I think this is mergeable at this point.

File: firewood/src/manager.rs (hunk 9 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs]: Will undo this diff + defer this to a separate issue: https://github.com/ava-labs/firewood/issues/1459

File: firewood/src/manager.rs (hunk 10 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs]: I learned this the hard way 😄(ref: aa64746). Fortunately, our tests failed immediately when I had this logic.

Fixed this here: 5bb0f67

File: firewood/src/manager.rs (hunk 9 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs]: Reverted diff here: d848744

File: firewood/src/manager.rs (hunk 11 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs:340]: Good call; done: f27fc4a

File: firewood/src/manager.rs (hunk 8 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs]: To me, reaping means adding the nodes of the revision to the free list - our current code seems to support this definition: 

https://github.com/ava-labs/firewood/blob/8ad5e19b56487388ea66e86d1726e3c8999a3543/firewood/src/manager.rs#L214-L216

This is different than rolling out the revision from memory, which is what I think you're referring to. I've updated the documentation to make this clear: ccca265 - feel free to resolve if the documentation for `commit()` is clear now.

File: ffi/firewood_test.go [hunk 13]
```diff
@@ -1244,6 +1244,58 @@ func TestHandlesFreeImplicitly(t *testing.T) {
...
+	for i := range numRevisions {
+		revision, err := db.Revision(revisionRoots[i])
+		r.NoError(err)
+
+		v, err := revision.Get(key)
+		r.NoError(err)
+
+		r.Equal(vals[i], v)
+		r.NoError(revision.Drop())
```
Comment (alarso16) [inline ffi/firewood_test.go:1311]: You should do this in a defer above, imo. This way, you won't get an error closing the database

File: firewood/src/manager.rs (hunk 12 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs]: Thanks for the suggestion. As discussed offline, I went with using a weak table to cache the revisions. Once the strong count of a weak pointer becomes 0, the key-value pair will be (lazily) removed: 18e6f13

File: ffi/firewood_test.go (hunk 2 above)
Comment (RodrigoVillar) [reply ffi/firewood_test.go:1298]: @rkuris Now that we're on the same page when it comes to the concept of reaping, could you revisit your comment above and update it if needed?

File: ffi/firewood_test.go (hunk 13 above)
Comment (RodrigoVillar) [reply ffi/firewood_test.go:1311]: Done: 5665a09

File: ffi/firewood_test.go (hunk 2 above)
Comment (rkuris) [reply ffi/firewood_test.go:1298]: Yes, I think you'll need to know whether the RootStore implementation requires keeping the old roots around. This isn't something that should not be owned by the RevisionManager, since they can get out of sync or not be initialized properly, it should be owned by the RootStore.

You can do that easily with a trait associated constant, maybe named ALLOW_SPACE_REUSE or something:

```rust
trait RootStore {
    const ALLOWSPACE_SPACE_REUSE: bool;
    ...
}
```

Then, each implementation of `RootStore` will require defining the value of that const.

File: firewood/src/manager.rs [hunk 14]
```diff
@@ -338,15 +340,56 @@ impl RevisionManager {
...
+
+        let nodestore = Arc::new(NodeStore::with_root(
+            root_hash.clone().into_hash_type(),
+            root_address,
+            self.current_revision(),
+        ));
+
+        // Cache the nodestore (stored as a weak reference).
+        self.by_rootstore
+            .lock()
```
Comment (rkuris) [inline firewood/src/manager.rs]: There's a race here (not a bug, but it wastes time if two callers hit here at the same time).

Reuse the lock from above; earlier, do:

```
let cache_guard = self.by_rootstore.lock().expect(...);
```

Then, leave `cache_guard` around and reuse it both earlier and here.

**Review by rkuris — CHANGES_REQUESTED** [review]
I'm not sure I love the separation of concerns here, but I could be convinced it's okay.

Perhaps the Rootstore interface should be that it returns nodestores themselves, and then the RootStore can deal with its own caching (only in the real implementation).

I think that cleans up the manager itself to focus on doing what it's supposed to be doing (tracking nodestores) instead of constructing them.

File: firewood/src/manager.rs (hunk 14 above)
Comment (RodrigoVillar) [reply firewood/src/manager.rs]: Done: a6cfff6

**Comment by RodrigoVillar** [comment]
@rkuis I'm definitely on-board with having `RootStore` own constructing and maintaining its own nodestores. With our current code, refactoring would fail due to the following circular dependency:

```mermaid
stateDiagram-v2
    s1: Revision Manager
    s2: RootStore

    s1 --> s2: Wants historical revision X
    s2 --> s1: Wants latest revision
```

One way to fix this would be to pass in the latest revision to `get()`:

```rust
fn get(
        &self,
        hash: &TrieHash,
        latest_revision: CommittedRevision,
    ) -> Result<Option<LinearAddress>, Box<dyn std::error::Error + Send + Sync>>;
```

However, passing in the latest revision seems odd (we expect `RootStore` to handle it's own nodestores). I think the better alternative would be to delegate the `historical` and `by_hash` fields of the `RevisionManager` into their own DS and have that be shared by both the `RevisionManager` and `RootStore`:

```rust
pub struct InMemoryRevisions {
    historical: VecDeque<CommittedRevision>,
    by_hash: HashMap<TrieHash, CommittedRevision>
}
```

Refactoring our code to use this DS would lead to the following dependency graph:

```mermaid
stateDiagram-v2
    s1: Revision Manager
    s2: RootStore
    s3: InMemoryRevisions

    s1 --> s3: Maintains latest revisions
    s1 --> s2: Wants historical revision
    s2 --> s3: Wants latest revision
```

File: ffi/firewood_test.go (hunk 2 above)
Comment (RodrigoVillar) [reply ffi/firewood_test.go:1298]: In the world of `dyn`'s, trait associated constants cannot be used unfortunately [[ref](https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility)]. 

I went ahead and added `allow_space_reuse()` as a method here: d20cd81. This should still be much more performant than having to call `get()` every time we commit.

**Comment by rkuris** [comment]
> ```rust
> pub struct InMemoryRevisions {
>     historical: VecDeque<CommittedRevision>,
>     by_hash: HashMap<TrieHash, CommittedRevision>
> }
> ```
Let's enter a new issue and do this. We should also have methods on this struct to manage and fetch from the list.

**Comment by RodrigoVillar** [comment]
Created #1470 which will track this.
